src/protocol/command.rs

```diff
@@ -1,10 +1,95 @@
 use super::argument_parser::Parser;
 use super::executable_path;
 use crate::R;
+use regex::Regex;
 use std::ffi::OsString;
 use std::path::PathBuf;
 use std::str;
 
+#[derive(Debug, Clone)]
+pub enum CommandMatcher {
+    Exact(Command),
+    RegexMatch(Regex),
+}
+
+impl PartialEq for CommandMatcher {
+    fn eq(&self, other: &CommandMatcher) -> bool {
+        use CommandMatcher::*;
+        match (self, other) {
+            (Exact(a), Exact(b)) => a == b,
+            (RegexMatch(a), RegexMatch(b)) => a.as_str() == b.as_str(),
+            _ => false,
+        }
+    }
+}
+
+impl Eq for CommandMatcher {}
+
+impl CommandMatcher {
+    pub fn matches_received(&self, received: &Command) -> bool {
```

Comment:
I think it's confusing that this has `received` in its name. I think from the perspective of the `CommandMatcher` it doesn't matter that this is used to match commands that are received. I think this should just be `matches` or `matches_command`.